### PR TITLE
Timelist bug fixes

### LINKED ISSUE #6536: Time List not scrolling to keep view centered on "now"
#### Summary
The Time List view is automatically scrolling its view to the latest possible activity in the view.
Screen Shot 2023-03-30 at 3 40 10 PM

#### Expected vs Current Behavior
The Time List view is supposed to scroll to keep activities that are close to "now" centered in the available view. This is key to making this view useful in a real-time ops context.
Screen Shot 2023-03-30 at 3 39 45 PM

#### Steps to Reproduce

1. Set the Time Conductor to a real-time mode.
2. View a Time List that includes a plan with activities in the past and the future based on settings from step 1.
3. Size the view such that the container displays a scroll bar. If needed, add the view to a layout.
4. Observe that the Time List scrolls to the list's extent and doesn't center the view.

#### Environment

* Open MCT Version:  2.2.1
* Deployment Type:  /testathon
* OS: Mac
* Browser: Chrome

#### Impact Check List

- [ ] Data loss or misrepresented data?
- [x] Regression? Did this used to work or has it always been broken?
- [ ] Is there a workaround available?
- [x] Does this impact a critical component?
- [ ] Is this just a visual bug with no functional impact?
- [ ] Does this block the execution of e2e tests?
- [ ] Does this have an impact on Performance?

### LINKED ISSUE #6657: Timelist To/From incorrect for 'current' activities
#### Summary
When an activity is 'current' (current timestamp is between the activity's start and end times) in a Timelist view, the To/From column should be displaying the `time to the end timestamp` of the activity.

#### Expected vs Current Behavior
Currently it shows the `time passed since the start timestamp`.

#### Steps to Reproduce
1. Create a timelist and drag a plan into it with at least 1 activity displaying as 'Current'.
2. Note the To/From column duration is incorrect.

#### Environment

* Open MCT Version: 2.2.2
* Deployment Type: All
* OS: All
* Browser: All

#### Impact Check List

- [x] Data loss or misrepresented data?
- [ ] Regression? Did this used to work or has it always been broken?
- [ ] Is there a workaround available?
- [ ] Does this impact a critical component?
- [ ] Is this just a visual bug with no functional impact?
- [ ] Does this block the execution of e2e tests?
- [ ] Does this have an impact on Performance?

### LINKED ISSUE #6677: Time List view should not show milliseconds in times
#### Summary
The current implementation of Time List includes milliseconds in timestamps and durations. This is not inline with the original design; this uses up needed space and makes times harder to parse.

#### Expected vs Current Behavior
Times should be shown without milliseconds.

#### Environment

* Open MCT Version: 
* Deployment Type: 
* OS:
* Browser:

#### Impact Check List

- [ ] Data loss or misrepresented data?
- [ ] Regression? Did this used to work or has it always been broken?
- [ ] Is there a workaround available?
- [ ] Does this impact a critical component?
- [x] Is this just a visual bug with no functional impact?
- [ ] Does this block the execution of e2e tests?
- [ ] Does this have an impact on Performance?

### PR DESCRIPTION
**Pull request by shefalijoshi**
Closes #6536
Closes #6657 
Closes #6677 

### Describe your changes:
Ensure timelist scrolling happens correctly for clock as well as fixed time - sort the list first, and wait for the DOM to be updated before scrolling.
If an activity has already started, then the duration should show the time to/since the end time of the activity.

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [x] Is this change backwards compatible? For example, developers won't need to change how they are calling the API or how they've extended core plugins such as Tables or Plots.

### Author Checklist

* [x] Changes address original issue?
* [ ] Tests included and/or updated with changes?
* [x] Command line build passes?
* [x] Has this been smoke tested?
* [x] Testing instructions included in associated issue OR is this a dependency/testcase change?

### Reviewer Checklist

* [ ] Changes appear to address issue?
* [ ] Reviewer has tested changes by following the provided instructions?
* [x] Changes appear not to be breaking changes?
* [ ] Appropriate automated tests included?
* [x] Code style and in-line documentation are appropriate?
* [x] Has associated issue been labelled unverified? (only applicable if this PR closes the issue)
* [x] Has associated issue been labelled bug? (only applicable if this PR is for a bug fix)

### CODE REVIEW

**Review by ozyx — CHANGES_REQUESTED** [review]
I've left a couple small suggestions. I have a couple of questions as well, which need @charlesh88 's input:

1. 
image

Looking at the above screenshot, you can see one activity that has passed its "end" time already, two which are currently active, and the rest which are yet to reach their "start" time. The time displayed for activities which have yet to occur is prefixed with a `+`, whereas activities which have already completed are prefixed with a `-`.

When I think of countdowns, I think T-Minus and T-Plus. This seems to reverse that (of course, I'm a layman with regards to rocketry so I could be totally wrong here). Is this the intended design?

2. 

https://github.com/nasa/openmct/assets/9259993/25292dfb-40c8-49a4-845d-f00424d836fd

Here I am navigating to a Timelist with a Plan that has two active activities. Upon navigation, the autoscroll will scroll down so that the first activity is out of view. The [bug description](https://github.com/nasa/openmct/issues/6536) states that "The Time List view is supposed to scroll to keep activities that are close to "now" centered in the available view." Does this mean that we should be scrolling so that the earliest active activity appears in the view, instead of the current behavior?

**Comment by charlesh88** [comment]
> When I think of countdowns, I think T-Minus and T-Plus. This seems to reverse that (of course, I'm a layman with regards to rocketry so I could be totally wrong here). Is this the intended design?

This appears to be the opposite of the original design: https://www.figma.com/file/oiWdYndFhxxkBTGkM9vWAv/Timelist-View-v1.0?type=design&node-id=0%3A1&t=fFwvfgmQBSyvY5uK-1. Basically, it's supposed to be like this:

- "Past" activities (start time < end time < now) should count UP from their END time with a "+" symbol.
- Current activities (start time < now < end time) should be counting DOWN to their END time, with a "-".
- Future activities (now < start time < end time) should be counting DOWN to their START time, with a "-".

The real thing is this: when the number are going up, it should be a "+" sign. When they're going down, they should use a "-".

**Comment by charlesh88** [comment]
> Here I am navigating to a Timelist with a Plan that has two active activities. Upon navigation, the autoscroll will scroll down so that the first activity is out of view. The [bug description](https://github.com/nasa/openmct/issues/6536) states that "The Time List view is supposed to scroll to keep activities that are close to "now" centered in the available view." Does this mean that we should be scrolling so that the earliest active activity appears in the view, instead of the current behavior?

I don't understand from your video why the second scroll event occurs. What's supposed to happen is that the view should scroll to keep current activities "centered" in the vertical space of the scrolled area. That's what appears to happen at first, but then a scroll occurs that hides the current activity. If that's what happened, that's not what's intended.

Can we do a screenshare.

**Comment by ozyx** [comment]
After discussing with Charles it looks like https://github.com/nasa/openmct/issues/6657 with regards to the timestamp is correctly fixed, although the `+` and the `-` signs need to be reversed (e.g. `+` prefixed timestamps should be counting down and `-` prefixed timestamps should be counting up). 
One other change @charlesh88 requested is to remove the milliseconds from the timestamps in the timelist, which should be achievable with just an edit to the moment format string.

There is an issue with the fix for https://github.com/nasa/openmct/issues/6536 in Display Layouts, where it seems that the scroll is not taking into account the frame size of the container, causing it to scroll where active activities may be out of view if embedded in a layout.

File: src/plugins/timelist/Timelist.vue [hunk 1]
```diff
@@ -346,10 +348,12 @@ export default {
                 activities = activities.concat(this.planData[key]);
             });
             activities = activities.filter(this.filterActivities);
+            activities = activities.sort(this.sortByStartTime);
             activities = this.applyStyles(activities);
-            this.setScrollTop();
             // sort by start time
             this.planActivities = activities.sort(this.sortByStartTime);
```
[the remark this answers is not in the document]
Comment (shefalijoshi) [reply src/plugins/timelist/Timelist.vue:354]: Yep. Removing.

**Comment by shefalijoshi** [comment]
> I've left a couple small suggestions. I have a couple of questions as well, which need @charlesh88 's input:
> 
> 
> image
> 
> Looking at the above screenshot, you can see one activity that has passed its "end" time already, two which are currently active, and the rest which are yet to reach their "start" time. The time displayed for activities which have yet to occur is prefixed with a `+`, whereas activities which have already completed are prefixed with a `-`.
> 
> When I think of countdowns, I think T-Minus and T-Plus. This seems to reverse that (of course, I'm a layman with regards to rocketry so I could be totally wrong here). Is this the intended design?
> 
> 
> Screen.Recording.2023-05-16.at.9.28.26.AM.mov
> 
> Here I am navigating to a Timelist with a Plan that has two active activities. Upon navigation, the autoscroll will scroll down so that the first activity is out of view. The [bug description](https://github.com/nasa/openmct/issues/6536) states that "The Time List view is supposed to scroll to keep activities that are close to "now" centered in the available view." Does this mean that we should be scrolling so that the earliest active activity appears in the view, instead of the current behavior?

@ozyx It looks like the current activity scrolls into view, yes? If so, this is intended behavior. If you manually scroll it, it should bounce back to the *current* activity after 1 minute of idling.

**Review by ozyx — APPROVED** [review]
This is looking good! I'm still seeing the issue regarding #6536 where the time list will scroll some current activities out of view if in a display layout (see quote below for more details), but in the interest of getting these changes in and in front of other eyes I'm going to approve.

The PR just needs a `lint:fix` and a small change against a failing unit test before it can merge

>  Screen.Recording.2023-05-16.at.9.28.26.AM.mov 
> Here I am navigating to a Timelist with a Plan that has two active activities. Upon navigation, the autoscroll will scroll down so that the first activity is out of view. The [bug description](https://github.com/nasa/openmct/issues/6536) states that "The Time List view is supposed to scroll to keep activities that are close to "now" centered in the available view." Does this mean that we should be scrolling so that the earliest active activity appears in the view, instead of the current behavior?

**Review by unlikelyzero — CHANGES_REQUESTED** [review]
Let's sync on the test coverage we want to add this this one. This issue is being track by Masoud

**Review by charlesh88 — DISMISSED** [review]
Fixes for no msecs and to/from/durations all look great. The fix for #6536 does not appear to be working, however, when there's no current activity. The view needs to center on "now" even when an activity isn't current.

**Current activity available, view centering properly**
Screen Shot 2023-06-05 at 11 16 59 AM

**No current activity - what the view _should_ be doing but isn't**
Screen Shot 2023-06-05 at 10 58 24 AM


**No current activity, view scrolls to top automatically**
Screen Shot 2023-06-05 at 10 58 14 AM

Synthetic plan used for testing:
[Plan Gen 2023-06-05.1.json.zip](https://github.com/nasa/openmct/files/11655639/Plan.Gen.2023-06-05.1.json.zip)

**Comment by shefalijoshi** [comment]
@charlesh88 Please re-review the fix for the final issue mentioned here.

**Comment by rukmini-bose** [comment]
Just reviewed this. The last change looks like its been fixed!

**Review by rukmini-bose — APPROVED** [review]
This looks good! Last change looks like it's been fixed.

**Review by unlikelyzero — APPROVED** [review]

**Pull request merged**